server/src/instant/grouped_queue.clj

```diff
@@ -0,0 +1,83 @@
+(ns instant.grouped-queue
+  (:refer-clojure :exclude [peek])
+  (:import
+   (java.util.concurrent LinkedBlockingQueue TimeUnit)
+   (java.util.concurrent.atomic AtomicInteger)
+   (clojure.lang PersistentQueue)))
+
+(def empty-q PersistentQueue/EMPTY)
+
+(defn create [{:keys [group-fn]}]
+  {:size (AtomicInteger. 0)
+   :group-fn group-fn
+   :group-key->subqueue (atom {})
+   :dispatch-queue (LinkedBlockingQueue.)})
+
+(defn size [{:keys [size] :as _grouped-q}]
+  (.get size))
+
+(defn put! [{:keys [group-fn dispatch-queue group-key->subqueue size]
+             :as _grouped-q} item]
+  (let [group-key (group-fn item)]
+    (if (nil? group-key)
+      ;; This item is not to be grouped.
+      (do (.incrementAndGet size)
+          (.put dispatch-queue [:item item]))
+
+      ;; This item will be grouped on `group-key` 
+      (let [_ (.incrementAndGet size)
+            [prev] (locking group-key->subqueue
+                     (swap-vals! group-key->subqueue
+                                 update
+                                 group-key
+                                 (fnil conj empty-q)
+                                 item))
+            prev-subqueue (get prev group-key)
+            first-enqueue? (empty? prev-subqueue)]
+        (when first-enqueue?
+          (.put dispatch-queue [:group-key group-key]))))))
+
+(defn peek [{:keys [dispatch-queue group-key->subqueue] :as _grouped-q}]
+  (let [[t arg :as entry] (.peek dispatch-queue)]
+    (cond
+      (nil? entry) nil
+      (= t :item) arg
+      (= t :group-key) (first (get @group-key->subqueue arg)))))
+
+(defn process-polling!
+  ([gq process-fn] (process-polling! gq process-fn {:poll-ms 1000}))
+  ([{:keys [dispatch-queue group-key->subqueue size] :as _grouped-q}
+    process-fn
+    {:keys [poll-ms]}]
+   (let [[t arg :as entry] (.poll dispatch-queue poll-ms TimeUnit/MILLISECONDS)]
+     (cond
+       (nil? entry) nil
+
+       (= t :item)
+       (do
+         (process-fn arg)
+         (.decrementAndGet size)
+         true)
+
+       (= t :group-key)
+       (let [group-key arg
+             item (first (get @group-key->subqueue group-key))]
```

Comment:
Just an idea, not sure I would implement it, but you might have a place in the queue object for the item/items that you're operating on. That way you could take it off the queue and it might make it easier to combine pending items, since you'd know you can combine anything in the queue.

It might look like:

```clojure
queue = {pending: PersistentQueue[item]
         working: [item]}
 ```
 
 And then the get would look something like:
 
 ```clojure
(swap! 
  big-queue
  key 
  (fn [] 
    (-> queue 
      (update :pending (.pop)
      (assoc :working [(first (:pending queue]))])
 ```
 
 And the complete would just clear the `:working` array.

Reply:
Oo. I like this idea; it adds more introspectability too. 

I can try this in a follow-on PR (will do shortly). Maybe we can already start batching!